### ISSUE TITLE
Bug: [no-unnecessary-condition] typed array creates false positive

### ISSUE THREAD

**Issue by plumpNation** [typescript-eslint/typescript-eslint#9543]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.3&showAST=types&fileType=.tsx&code=MYewdgzgLgBGCuBbARgUwE4QFxyW9A2gLowC8MxA3AFDWiSwSoA2qwUqAJgHJ4Zm4UGCAQCMABkmSiNagHo5MbiBgZ0IdHXAQQrAHTMQAcwAUTVuy68h6GAB87uZswCUshTAVfvP33-8BgV60qAAeAA4asACWYBzoAGYAhsCoMACSHIgwAN4wAL609NAw0VnYGVnEAlRaDDDmbBycmajZ5GVtIhJS4jK0HgCi6OroOACqYGBsqBAQSegAnjD0nGXR4EnMADQwAG5b8GnREDBbAO5Ji6dQ6PBQABaLdTr6hqaNli1Z9o4Izm4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const numbers: number[] = [];

const selectedNumber = numbers[100000];

// No error
console.log(selectedNumber || null);

// /////////////////////////////////////////////////////////
// The same behaviour is demonstrated with `Type` and `Interface`
export interface Item {}

const items: Item[] = [];
const selectedItem = items[100000];

// Error: Unnecessary conditional, value is always truthy
console.log(selectedItem || null);
```

### ESLint Config

```javascript
import tseslint from 'typescript-eslint';

export default tseslint.config(
  ...tseslint.configs.strictTypeChecked,
  {
    languageOptions: {
      parserOptions: {
        project: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
  },
);
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": "true"
  }
}
```

### Expected Result

I expect that `selectedItem` should be potentially undefined. Seeing as that is not the function of this rule, I would expect that the rule does not create a "Unnecessary conditional, value is always truthy" error for `selectedItem || null`.

### Actual Result

"Unnecessary conditional, value is always truthy" is seen for `selectedItem || null`.

### Additional Info

```
// package.json
{
  "devDependencies": {
    "@eslint/js": "9.6.0",
    "@types/eslint__js": "8.42.3",
    "eslint": "9.6.0",
    "typescript": "5.5.3",
    "typescript-eslint": "7.16.0"
  }
}
```

The config in the playground uses the eslintrc format. The config in this issue is the one used in my test repo. The result is the same.

Replicated in intellij, terminal, playground, and vscode.

**Comment by bradzacher** [comment]
The rule is powered by types and the type of `selectedItem` is non-nullable.

See the following FAQ:
https://typescript-eslint.io/troubleshooting/typed-linting#typescript-eslint-thinks-my-variable-is-never-nullish--is-any--etc-but-that-is-clearly-not-the-case-to-me

you probably want something like https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess

**Comment by plumpNation** [comment]
`noUncheckedIndexedAccess` does indeed fix this issue. Apologies for wasting your time with this, and thank you very much for the tips and the tool. It is greatly appreciated.
